File: objective-c-sample/SendBirdSampleUI/SendBird/SendBirdMessaging/View/TypingNowView.m
//
//  TypingNowView.m
//  SendBirdSampleUI
//
//  Created by SendBird Developers on 2015. 7. 29..
//  Copyright (c) 2015 SENDBIRD.COM. All rights reserved.
//

#import "TypingNowView.h"

@implementation TypingNowView

- (id) initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self != nil) {
        [self initView];
    }
    return self;
}

- (void) initView
{
    [self setBackgroundColor:[UIColor whiteColor]];
    
    self.typingImageView = [[UIImageView alloc] init];
    [self.typingImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    self.typingImageView.animationImages = [NSArray arrayWithObjects:
                                            [UIImage imageNamed:@"_icon_loading_typing0"],
                                            [UIImage imageNamed:@"_icon_loading_typing1"],
                                            [UIImage imageNamed:@"_icon_loading_typing2"],
                                            [UIImage imageNamed:@"_icon_loading_typing3"],
                                            [UIImage imageNamed:@"_icon_loading_typing4"],
                                            [UIImage imageNamed:@"_icon_loading_typing5"],
                                            [UIImage imageNamed:@"_icon_loading_typing6"], nil];
    
    self.typingImageView.animationDuration = 1.0f;
    self.typingImageView.animationRepeatCount = 0;
    [self.typingImageView startAnimating];
    [self addSubview:self.typingImageView];
    
    self.typingLabel = [[UILabel alloc] init];
    [self.typingLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.typingLabel setFont:[UIFont italicSystemFontOfSize:12.0]];
    [self.typingLabel setNumberOfLines:0];
    [self.typingLabel setTextColor:UIColorFromRGB(0x9d9ba5)];
    [self.typingLabel setText:@"Typing something cool...."];
    [self.typingLabel setLineBreakMode:NSLineBreakByCharWrapping];
    [self addSubview:self.typingLabel];
    
    // Typing Image View
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.typingImageView
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:10]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.typingImageView
                                                     attribute:NSLayoutAttributeCenterY
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.typingLabel
                                                     attribute:NSLayoutAttributeCenterY
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.typingImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.typingLabel
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:-10]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.typingImageView
                                                     attribute:NSLayoutAttributeWidth
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:18.5]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.typingImageView
                                                     attribute:NSLayoutAttributeHeight
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:5]];
    
    // Typing Label
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.typingLabel
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:-8]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.typingLabel
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:10]];
}

- (void) setModel:(NSDictionary *)typeStatus
{
    int count = (int)[typeStatus count];
    [self.typingLabel setText:[NSString stringWithFormat:@"%d Typing something cool....", count]];
}

@end

File: swift-sample/SendBirdSampleUI/SendBird/SendBirdMessaging/View/TypingNowView.swift
//
//  TypingNowView.swift
//  SendBirdSampleUI
//
//  Created by Jed Kyung on 2/7/16.
//  Copyright © 2016 SENDBIRD.COM. All rights reserved.
//

import UIKit
import SendBirdSDK

class TypingNowView: UIView {
    var typingImageView: UIImageView?
    var typingLabel: UILabel?

    convenience init() {
        self.init(frame:CGRect.zero)
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.initView()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func initView() {
        self.backgroundColor = UIColor.whiteColor()
        
        self.typingImageView = UIImageView()
        self.typingImageView?.translatesAutoresizingMaskIntoConstraints = false
        self.typingImageView?.animationImages = [UIImage.init(named: "_icon_loading_typing0")!,
            UIImage.init(named: "_icon_loading_typing1")!,
            UIImage.init(named: "_icon_loading_typing2")!,
            UIImage.init(named: "_icon_loading_typing3")!,
            UIImage.init(named: "_icon_loading_typing4")!,
            UIImage.init(named: "_icon_loading_typing5")!,
            UIImage.init(named: "_icon_loading_typing6")!,
        ]
        self.typingImageView?.animationDuration = 1.0
        self.typingImageView?.animationRepeatCount = 0
        self.typingImageView?.startAnimating()
        self.addSubview(self.typingImageView!)
        
        self.typingLabel = UILabel()
        self.typingLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.typingLabel?.font = UIFont.italicSystemFontOfSize(12.0)
        self.typingLabel?.numberOfLines = 0
        self.typingLabel?.textColor = SendBirdUtils.UIColorFromRGB(0x9d9ba5)
        self.typingLabel?.text = "Typing something cool..."
        self.typingLabel?.lineBreakMode = NSLineBreakMode.ByCharWrapping
        self.addSubview(self.typingLabel!)
        
        // Typing Image View
        self.addConstraint(NSLayoutConstraint.init(item: self.typingImageView!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: 10))
        self.addConstraint(NSLayoutConstraint.init(item: self.typingImageView!, attribute: NSLayoutAttribute.CenterY, relatedBy: NSLayoutRelation.Equal, toItem: self.typingLabel!, attribute: NSLayoutAttribute.CenterY, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.typingImageView!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self.typingLabel!, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: -10))
        self.addConstraint(NSLayoutConstraint.init(item: self.typingImageView!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 18.5))
        self.addConstraint(NSLayoutConstraint.init(item: self.typingImageView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 5))
        
        // Typing Label
        self.addConstraint(NSLayoutConstraint.init(item: self.typingLabel!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: -8))
        self.addConstraint(NSLayoutConstraint.init(item: self.typingLabel!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: 10))
    }
    
    func setModel(typeStatus: NSDictionary) {
        let count: Int = typeStatus.count
        self.typingLabel?.text = String.init(format: "%d Typing something cool...", count)
    }
}
